docs/tutorial/write-a-machine-charm-for-tinyproxy.md

```diff
@@ -0,0 +1,980 @@
+```
+
+Then add the following method to the charm class:
+
+```python
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        """Report the status of tinyproxy (runs after each event)."""
+        try:
+            self.load_config(TinyproxyConfig)
+        except pydantic.ValidationError as e:
+            slug_error = e.errors()[0]  # Index 0 because 'slug' is the only option validated.
+            slug_value = slug_error["input"]
+            message = f"Invalid slug: '{slug_value}'. Slug must match the regex [a-z0-9-]+"
+            event.add_status(ops.BlockedStatus(message))
+        if not tinyproxy.is_installed():
+            event.add_status(ops.MaintenanceStatus("Waiting for tinyproxy to be installed"))
+        if not tinyproxy.is_running():
+            event.add_status(ops.MaintenanceStatus("Waiting for tinyproxy to start"))
+        event.add_status(ops.ActiveStatus())
+```
+
+Ops runs this method after each Juju event, regardless of whether the charm code handles the event. After running the method, Ops decides which status to report to Juju, choosing the highest priority status that was proposed with `event.add_status`.
+
+For example, if the value of the `slug` configuration option is invalid, `load_config` raises an error and the charm code proposes a "blocked" status. This status means that your charm needs intervention from a human, so Ops gives it the highest priority and reports it to Juju.
+
+### Stop and uninstall tinyproxy
+
+If Juju wants to remove a unit of the application, your charm (on that unit) receives the "stop" event followed by the "remove" event.
+
+To handle these events, add the following lines to the `__init__` method of the charm class:
+
+```python
+        framework.observe(self.on.stop, self._on_stop)
+        framework.observe(self.on.remove, self._on_remove)
+```
+
+Then add the following methods to the charm class:
+
+```python
+    def _on_stop(self, event: ops.StopEvent) -> None:
+        """Handle stop event."""
+        tinyproxy.stop()
+        self.wait_for_not_running()
+
+    def _on_remove(self, event: ops.RemoveEvent) -> None:
+        """Handle remove event."""
+        tinyproxy.uninstall()
+
+    def wait_for_not_running(self) -> None:
+        """Wait for tinyproxy to not be running."""
+        for _ in range(3):
+            if not tinyproxy.is_running():
+                return
+            time.sleep(1)
+        raise RuntimeError("tinyproxy was still running after the expected time")
+```
+
+## Try your charm
+
+### Pack your charm
+
+Before you can try your charm, you need to "pack" it. Packing combines the charm code and metadata into a single file that can be deployed to Juju.
+
+In your virtual machine, make sure that the working directory is `~/tinyproxy`. Then run:
+
+```text
+charmcraft pack
+```
+
+Charmcraft will take about 20 minutes to pack your charm, depending on your computer and network.
```

Comment:
Since we're using the `uv` plugin and exploiting the bug to avoid building `pydantic`, this seems excessive. I do find that it's regularly about 20 minutes if you're building everything from source, but this took 8m35s for me because we skipped that. Maybe we should be less certain about this and say it could take as much as 20 minutes, or something like that?

Reply:
Yeah, this took 2m29s for me.